gk/main.c

```diff
@@ -552,6 +552,71 @@ gk_setup_rss(struct gk_config *gk_conf)
 	return ret;
 }
 
+static void
+gk_arp_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	unsigned int i;
+	struct gk_lpm *ltbl = arg;
+
+	for (i = 0; i < ltbl->num_tunnels; i++) {
+		if (ltbl->tunnels[i].flow.proto == ETHER_TYPE_IPv4
+				&& memcmp(&ltbl->tunnels[i].flow.f.v4.dst,
+				map->ip_be, sizeof(struct in_addr)) == 0) {
+			ether_addr_copy(&map->ha,
+				&ltbl->tunnels[i].nexthop_mac);
```

Comment:
Sequential locks are enough for this need; see Chapter 10 (page 200) of the Linux Kernel Development book by Robert Love. Another option is to embrace liburcu (http://liburcu.org/) and adopt RCU in Gatekeeper. Although I like RCU, I think sequential locks are all we need here.

A side note, thinking of this problem I found another issue in @mengxiang0811's code: it ignores the stale flag. If an entry is stale, GK blocks must drop the outgoing packet instead of sending them out with stale, or even invalid, destination MAC addresses.